hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -570,24 +570,30 @@ def each_pair_of_blocks(self, accept_first, accept_second):
         """Yield each pair of blocks ``(a, b)``, such that ``a.index <
         b.index``, but only if ``accept_first(a)`` and ``accept_second(b)`` are
         both true."""
-        i = 0
-        while i < len(self.blocks):
-            j = i + 1
-            while j < len(self.blocks):
+
+        # Iteration order here is significant: Rather than fixing i and looping
+        # over each j, then doing the same, etc. we iterate over the gap between
+        # i and j and then over i. The reason for this is that it ensures that
+        # we try a different value for i and j on each iteration of the inner
+        # loop. This stops us from stalling if we happen to hit on a value of i
+        # where nothing useful can be done.
+        #
+        # In the event that nothing works, this doesn't help and we still make
+        # the same number of calls, but by ensuring that we make progress we
+        # have more opportunities to make shrinks that speed up the tests or
+        # that reduce the number of viable shrinks at the next gap size because
+        # we've lowered some values.
+        gap = 1
```

Comment:
Minor: The name `gap` is slightly awkward, because the number is more of an offset or delta than an actual gap.

(E.g. when `gap = 1`, `i` and `j` are adjacent, so they are offset by 1 but there is no gap between them.)

Reply:
Thinking about this again, I guess there is a case to be made that adjacent integers do have a “gap” between them. So this is more a question of perspective than right/wrong.

(Either way, the code itself is pretty clear, so leaving it as-is would be fine.)